README.md
# pubs-extract

[![status-badge](https://ci.martyoeh.me/api/badges/Marty/pubs-extract/status.svg)](https://ci.martyoeh.me/Marty/pubs-extract)
![PyPI](https://img.shields.io/pypi/v/pubs-extract)

Quickly extract annotations from your pdf files with the help of the pubs bibliography manager.
Easily organize your highlights and thoughts next to your documents.

## Installation:

You can install from pypi with `pip install pubs-extract`.

Or you install manually by moving the `extract` directory into your pubs `plugs` directory,
so that the hierarchy is `pubs/plugs/extract/`

Then add `extract` to your plugin list in the pubs configuration file:

```ini
[plugins]
active = extract
```

To check if everything is working you can do `pubs --help` which should show you the new extract command.
You will be set up with the default options but if you want to change anything, read on in configuration below.

> **Note**
> This plugin is in fairly early development. It does what I need it to do, but if you have a meticulously organized library *please* make backups before doing any operation on your notes, or make use of the pubs-included git plugin.

## Configuration:

In your pubs configuration file:

```ini
[plugins]
active = extract

[[extract]]
on_import = False
short_header = False
minimum_text_similarity = 0.75
minimum_color_similarity = 0.833
formatting = "{%quote_container> {quote} %}[{page}]{%note_container{newline}Note: {note} %}{%tag_container #{tag}%}"
```

If `on_import` is `True` extraction is automatically run whenever a new document is added to the library,
if false extraction has to be handled manually.

---

`short_header` determines if the headline of each annotation output (displaying the paper it is from) should contain the whole formatted author, year, title string (`False`) or just the citekey (`True`).

---

`minimum_text_similarity` sets the required similarity of an annotation's note and written words to be viewed
as one. Any annotation that has both and is *under* the minimum similarity will be added in the following form:

```markdown
> [13] my annotation
Note: my additional thoughts
```

That is, the extractor detects additional written words by whoever annotated and adds them to the extraction.
The option generally should not take too much tuning, but it is there if you need it.

---

`minimum_color_similarity` sets the required similarity of highlight/annotation colors to be recognized as the 'pure' versions of themselves for color mapping (see below). With a low required similarity, for example dark green and light green will both be recognized simply as 'green' while a high similarity will not match them, instead only matching closer matches to a pure (0, 255, 0) green value.

This should generally be an alright default but is here to be changed for example if you work with a lot of different annotation colors (where dark purple and light purple get different meanings) and get false positives.

---

The plugin contains a configuration sub-category of `tags`: Here, you can define meaning for your highlight/annotation colors. For example, if you always highlight the main arguments and findings in orange and always highlight things you have to follow up on in blue, you can assign the meanings 'important' and 'todo' to them respectively as follows:

```ini
[[[tags]]]
orange = "important"
blue = "todo"
```

Currently recognized colors are: `red` `green` `blue` `yellow` `purple` `orange`.
Since these meanings are often highly dependent on personal organization and reading systems,
no defaults are set here.

---

`formatting` takes a string with a variety of template options. You can use any of the following:

- `{page}`: The page number the annotation was found on.
- `{quote}`: The actual quoted string (i.e. highlighted).
- `{note}`: The annotation note (i.e. addded reader).
- `{%quote_container [other text] %}`: Mark the area that contains a quotation. Useful to get rid of prefixes or suffixes if no quotation exists. Usually contains some plain text and a `{quote}` element. Can *not* be nested with other containers.
- `{%note_container [other text] %}`: Mark the area that contains a note. Useful to get rid of prefixes or suffixes if no note exists. Usually contains some plain text and a `{note}` element. Can *not* be nested with other containers.
- `{%tag_container [other text] %}`: Mark the area that contains a tag. Useful to get rid of prefixes or suffixes if no tag exists. Usually contains some plain text and a `{tag}` element. Can *not* be nested with other containers.
- `{newline}`: Add a line break in the resulting annotation display.

For example, the default formatting string `"{%quote_container> {quote} %}[{page}]{%note_container{newline}Note: {note} %}{%tag_container #{tag}%}"` will result in this output:

```
> Mobilizing the TPSN scheme (see above) and drawing on cultural political economy and critical governance studies, this landmark article offers an alternative account [5]
Note: a really intersting take on polydolywhopp
```

Container marks are useful to encapsulate a specific type of the annotation, so extracted annotations in the end don't contains useless linebreaks or quotations markup.

## Usage:

`pubs extract [-h|-w|-e] <citekeys>`

For example, to extract annotations from two entries, do:

```bash
pubs extract Bayat2015 Peck2004
```

This will print the extracted annotations to the commandline through stdout.

If you invoke the command with the `-w` option, it will write it into your notes instead:

```bash
pubs extract -w Bayat2015 Peck2004
```

Will create notes for the two entries in your pubs note directory and fill them with
the annotations. If a note already exists for any of the entries, it will instead append
the annotations to the end of it, dropping all those that it already finds in the note
(essentially only adding new annotations to the end).

**PLEASE** Heed the note above and exercise caution with the `-w` option. It is
not intended to be destructive, but nevertheless create backups or version
control your files.

You can invoke the command with `-e` to instantly edit the notes in your editor:

```bash
pubs extract -w -e Bayat2015 Peck2004
```

Will create/append annotations and drop you into the Bayat2015 note, when you
close it directly into the Peck2004 note. Take care that it will be fairly
annoying if you use this option with hundreds of entries being annotated as it
will open one entry after another for editing.

To extract the annotations for all your existing entries in one go, you can use:

```bash
pubs extract -w $(pubs list -k)
```

However, the warning for your notes' safety goes doubly for this command since it will touch
*most* or *all* of your notes, depending on how many entries in your library have pdfs attached.

This readme is still a bit messy, feel free to extend it and raise a PR if you have the time.

What follows is a not-very-sorted train of though on where the plugin is at and where I
could see myself taking it one day, provided I find the time.
Pull requests tackling one of these areas of course very welcome.

## Issues

A note on the extraction: Highlights in pdfs can be somewhat difficult to parse
(as are most things in them). Sometimes they contain the selected text that is written on the
page, sometimes they contain the annotators thoughts as a note, sometimes they contain nothing.
This plugin makes an effort to find the right combination and extract the written words,
as well as any additional notes made - but things *will* slip through or extract weirdly every now
and again.

The easiest extraction is provided if your program writes the selection itself into the highlight
content, because then we can just use that. It is harder to parse if it does not and will sometimes
get additional words in front or behind (especially if the highlight ends in the middle of a line)
or even cut a few off.

I am not sure if there is much I can do about this.

---

If you spot a bug or have an idea feel free to open an issue.\
I might be slow to respond but will consider them all!

Pull requests are warmly welcomed.\
If they are big changes or additions let's talk about them in an issue first.

Thanks for using my software ❤️


CHANGELOG.md
# Changelog

All notable changes to this project will be documented in this file.

The format is based on [Keep a Changelog](https://keepachangelog.com/en/1.0.0/),
and this project tries to adhere to [Semantic Versioning](https://semver.org/spec/v2.0.0.html).

<!--
## [Unreleased]

### Added

### Changed

### Fixed
-->

## [0.2.0] - 2023-01-13

### Added

* Added pypi release publication
* Try not to duplicate extracted annotations in existing notes
* Map annotation colors to custom tags in notes
* Add querying for publications to command (like list command)

### Fixed

* Grab annotations even if their content is empty or contains custom text

## [0.1.0] - 2022-12-25

* Extract highlights and annotations from a pubs doc file
* Optionally run automatically whenever file is added to pubs
* Optionally write annotations to pubs note file


tools/extract-changelog.py
import re

## Extracts the version and newest changes from a semantic changelog.
#
# Important, it only works with three-parted version numbers
# a-la 1.2.3 or 313.01.1888 -- needs \d.\d.\d to work.
#
# The version number and changeset will be put in `NEWEST_VERSION.md`
# and `NEWEST_CHANGES.md` respectively, for further use in releases.
OUTPUT_FILE_VERSION = "NEWEST_VERSION.md"
OUTPUT_FILE_CHANGES = "NEWEST_CHANGES.md"


def getVersion(file):
    for line in file:
        m = re.match(r"^## \[(\d+\.\d+\.\d+)\]", line)
        if m and m.group(1):
            return m.group(1)


def getSection(file):
    inRecordingMode = False
    for line in file:
        if not inRecordingMode:
            if re.match(r"^## \[\d+\.\d+\.\d+\]", line):
                inRecordingMode = True
        elif re.match(r"^## \[\d+\.\d+\.\d+\]", line):
            inRecordingMode = False
            break
        elif re.match(r"^$", line):
            pass
        else:
            yield line


def toFile(fname, content):
    file = open(fname, "w")
    file.write(content)
    file.close()


with open("CHANGELOG.md") as file:
    title = getVersion(file)
    print(title)
    toFile(OUTPUT_FILE_VERSION, title)

with open("CHANGELOG.md") as file:
    newest_changes_gen = getSection(file)
    newest_changes = ""
    for line in newest_changes_gen:
        newest_changes += line
    print("[Extracted Changelog]")
    print(newest_changes)
    toFile(OUTPUT_FILE_CHANGES, newest_changes)

file.close()


pubs/plugs/extract/extract.py
import os
import argparse

import fitz
import Levenshtein

from pubs.plugins import PapersPlugin
from pubs.events import DocAddEvent, NoteEvent
from pubs import repo, pretty
from pubs.utils import resolve_citekey_list
from pubs.content import check_file, read_text_file, write_file
from pubs.query import get_paper_filter
from .annotation import (
    PaperAnnotated,
    Annotation,
    COLOR_SIMILARITY_MINIMUM,
    TEXT_SIMILARITY_MINIMUM,
)

CONFIRMATION_PAPER_THRESHOLD = 5


class ExtractPlugin(PapersPlugin):
    """Extract annotations from any pdf document.

    The extract plugin allows manual or automatic extraction of all annotations
    contained in the pdf documents belonging to entries of the pubs library.

    It can write those changes to stdout or directly create and update notes
    for the pubs entries.

    It adds a `pubs extract` subcommand through which it is invoked, but can
    optionally run whenever a new document is imported for a pubs entry.
    """

    name = "extract"
    description = "Extract annotations from pubs documents"

    def __init__(self, conf, ui):
        self.ui = ui
        self.note_extension = conf["main"]["note_extension"]
        self.max_authors = conf["main"]["max_authors"]
        self.repository = repo.Repository(conf)
        self.pubsdir = os.path.expanduser(conf["main"]["pubsdir"])
        self.broker = self.repository.databroker

        settings = conf["plugins"].get("extract", {})
        self.on_import = settings.get("on_import", False)
        self.minimum_similarity = float(
            settings.get("minimum_text_similarity", TEXT_SIMILARITY_MINIMUM)
        )
        self.minimum_color_similarity = float(
            settings.get("minimum_color_similarity", COLOR_SIMILARITY_MINIMUM)
        )
        self.formatting = settings.get(
            "formatting",
            "{%quote_container> {quote} %}[{page}]{%note_container{newline}Note: {note} %}{%tag_container #{tag}%}",
        )
        self.color_mapping = settings.get("tags", {})
        self.short_header = settings.get("short_header", False)

    def update_parser(self, subparsers, _):
        """Allow the usage of the pubs extract subcommand"""
        # TODO option for ignoring missing documents or erroring.
        extract_parser = subparsers.add_parser(self.name, help=self.description)
        extract_parser.add_argument(
            "-w",
            "--write",
            help="Write to individual notes instead of standard out. Appends to existing notes.",
            action="store_true",
            default=None,
        )
        extract_parser.add_argument(
            "-e",
            "--edit",
            help="Open each note in editor for manual editing after extracting annotations to it.",
            action="store_true",
            default=False,
        )
        extract_parser.add_argument(
            "-q",
            "--query",
            help="Query library instead of providing individual citekeys. For query help see pubs list command.",
            action="store_true",
            default=None,
            dest="is_query",
        )
        extract_parser.add_argument(
            "-i",
            "--ignore-case",
            action="store_false",
            default=None,
            dest="case_sensitive",
            help="When using query mode, perform case insensitive search.",
        )
        extract_parser.add_argument(
            "-I",
            "--force-case",
            action="store_true",
            dest="case_sensitive",
            help="When using query mode, perform case sensitive search.",
        )
        extract_parser.add_argument(
            "--strict",
            action="store_true",
            default=False,
            help="Force strict unicode comparison of query.",
        )
        extract_parser.add_argument(
            "query",
            nargs=argparse.REMAINDER,
            help="Citekey(s)/query for the documents to extract from.",
        )
        extract_parser.set_defaults(func=self.command)

    def command(self, conf, args):
        """Run the annotation extraction command."""
        papers = self._gather_papers(conf, args)
        all_annotations = self.extract(papers)
        if args.write:
            self._to_notes(all_annotations, self.note_extension, args.edit)
        else:
            self._to_stdout(all_annotations, self.short_header)
        self.repository.close()

    def extract(self, papers):
        """Extracts annotations from citekeys.

        Returns all annotations belonging to the papers that
        are described by the citekeys passed in.
        """
        papers_annotated = []
        for paper in papers:
            file = self._get_file(paper)
            try:
                annotations = self._get_annotations(file)
                papers_annotated.append(PaperAnnotated.from_paper(paper, annotations))
            except fitz.FileDataError as e:
                self.ui.error(f"Document {file} is broken: {e}")
        return papers_annotated

    def tag_from_colorname(self, colorname):
        return self.color_mapping.get(colorname, "")

    def _gather_papers(self, conf, args):
        """Get all papers for citekeys.

        Returns all Paper objects described by the citekeys
        passed in.
        """
        papers = []
        if not args.is_query:
            keys = resolve_citekey_list(
                self.repository, conf, args.query, ui=self.ui, exit_on_fail=True
            )
            if not keys:
                return []
            for key in keys:
                papers.append(self.repository.pull_paper(key))
        else:
            papers = list(
                filter(
                    get_paper_filter(
                        args.query,
                        case_sensitive=args.case_sensitive,
                        strict=args.strict,
                    ),
                    self.repository.all_papers(),
                )
            )
        if len(papers) > CONFIRMATION_PAPER_THRESHOLD:
            self.ui.message(
                "\n".join(
                    pretty.paper_oneliner(
                        p, citekey_only=False, max_authors=self.max_authors
                    )
                    for p in papers
                )
            )
            self.ui.input_yn(
                question=f"Extract annotations for these papers?", default="y"
            )
        return papers

    def _get_file(self, paper):
        """Get path of document belonging to paper.

        Returns the real path to the document which belongs
        to the paper passed in. Emits a warning if no
        document belongs to paper.
        """
        path = self.broker.real_docpath(paper.docpath)
        if not path:
            self.ui.warning(f"{paper.citekey} has no valid document.")
        return path

    def _get_annotations(self, filename):
        """Extract annotations from a file.

        Returns all readable annotations contained in the file
        passed in. Only returns Highlight or Text annotations
        currently.
        """
        annotations = []
        with fitz.Document(filename) as doc:
            for page in doc:
                for annot in page.annots():
                    quote, note = self._retrieve_annotation_content(page, annot)
                    a = Annotation(
                        file=filename,
                        text=quote,
                        content=note,
                        colors=annot.colors,
                        type=annot.type[1],
                        page=(page.number or 0) + 1,
                    )
                    a.tag = self.tag_from_colorname(a.colorname)
                    annotations.append(a)
        return annotations

    def _retrieve_annotation_content(self, page, annotation):
        """Gets the text content of an annotation.

        Returns the actual content of an annotation. Sometimes
        that is only the written words, sometimes that is only
        annotation notes, sometimes it is both. Runs a similarity
        comparison between strings to find out whether they
        should both be included or are doubling up, using
        Levenshtein distance.
        """
        content = annotation.info["content"].replace("\n", " ")
        written = page.get_textbox(annotation.rect).replace("\n", " ")

        # highlight with selection in note
        if Levenshtein.ratio(content, written) > self.minimum_similarity:
            return (content, "")
        # an independent note, not a highlight
        elif content and not written:
            return ("", content)
        # both a highlight and a note
        elif content:
            return (written, content)
        # highlight with selection not in note
        return (written, "")

    def _to_stdout(self, annotated_papers, short_header=False):
        """Write annotations to stdout.

        Simply outputs the gathered annotations over stdout
        ready to be passed on through pipelines etc.
        """
        output = ""
        for paper in annotated_papers:
            output += f"\n------ {paper.headline(self.short_header, self.max_authors)} ------\n\n"
            for annotation in paper.annotations:
                output += f"{annotation.format(self.formatting)}\n"
                output += "\n"
        self.ui.message(output.strip())

    def _to_notes(self, annotated_papers, note_extension="txt", edit=False):
        """Write annotations into pubs notes.

        Permanently writes the given annotations into notes
        in the pubs notes directory. Creates new notes for
        citekeys missing a note or appends to existing.
        """
        for paper in annotated_papers:
            if paper.annotations:
                notepath = self.broker.real_notepath(paper.citekey, note_extension)
                if check_file(notepath, fail=False):
                    self._append_to_note(notepath, paper)
                else:
                    self._write_new_note(
                        notepath,
                        paper,
                        paper.headline(short=True, max_authors=self.max_authors),
                    )
                self.ui.info(f"Wrote annotations to {paper.citekey} note {notepath}.")

                if edit is True:
                    self.ui.edit_file(notepath, temporary=False)
                NoteEvent(paper.citekey).send()

    def _write_new_note(self, notepath, paper, headline):
        """Create a new note containing the annotations.

        Will create a new note in the notes folder of pubs
        and fill it with the annotations extracted from pdf.
        """
        output = f"# {headline}\n\n"
        for annotation in paper.annotations:
            output += f"{annotation.format(self.formatting)}\n\n"
        write_file(notepath, output, "w")

    def _append_to_note(self, notepath, paper):
        """Append new annotations to the end of a note.

        Looks through note to determine any new annotations which should be
        added and adds them to the end of the note file.
        """
        existing = read_text_file(notepath)
        # removed annotations already found in the note
        existing_dropped = [
            x for x in paper.annotations if x.format(self.formatting) not in existing
        ]
        if not existing_dropped:
            return

        output = ""
        for annotation in existing_dropped:
            output += f"{annotation.format(self.formatting)}\n\n"
        write_file(notepath, output, "a")


@DocAddEvent.listen()
def modify_event(event):
    if ExtractPlugin.is_loaded():
        plg = ExtractPlugin.get_instance()
        if plg.on_import:
            all_annotations = plg.extract([event.citekey])
            if all_annotations[0][1]:
                plg._to_notes(all_annotations, plg.note_extension)


pubs/plugs/extract/annotation.py
import math
import re
from dataclasses import dataclass, field
from typing import Dict

from pubs.paper import Paper
from pubs import pretty

TEXT_SIMILARITY_MINIMUM = 0.75
COLOR_SIMILARITY_MINIMUM = 0.833

COLORS = {
    "red": (1, 0, 0),
    "green": (0, 1, 0),
    "blue": (0, 0, 1),
    "yellow": (1, 1, 0),
    "purple": (0.5, 0, 0.5),
    "orange": (1, 0.65, 0),
}


class PaperAnnotated(Paper):
    def __init__(self, citekey, bibdata, metadata=None, annotations=[]):
        super(PaperAnnotated, self).__init__(citekey, bibdata, metadata)
        self.annotations = annotations

    @classmethod
    def from_paper(cls, paper, annotations=[]):
        return cls(paper.citekey, paper.bibdata, paper.metadata, annotations)

    def __repr__(self):
        return "PaperAnnotated(%s, %s, %s)" % (
            self.citekey,
            self.bibdata,
            self.metadata,
        )

    def headline(self, short=False, max_authors=3):
        headline = pretty.paper_oneliner(
            self, citekey_only=short, max_authors=max_authors
        )
        return re.sub(r"\[pdf\]", "", headline).rstrip()


@dataclass
class Annotation:
    """A PDF annotation object"""

    file: str
    type: str = "Highlight"
    text: str = ""
    content: str = ""
    page: int = 1
    colors: Dict = field(default_factory=lambda: {"stroke": (0.0, 0.0, 0.0)})
    tag: str = ""

    def format(self, formatting):
        """Return a formatted string of the annotation.

        Given a provided formatting pattern, this method returns the annotation
        formatted with the correct marker replacements and removals, ready
        for display or writing.
        """
        output = formatting
        replacements = {
            r"{quote}": self.text,
            r"{note}": self.content,
            r"{page}": str(self.page),
            r"{newline}": "\n",
            r"{tag}": self.tag,
        }
        pattern = re.compile(
            "|".join(
                [re.escape(k) for k in sorted(replacements, key=len, reverse=True)]
            ),
            flags=re.DOTALL,
        )
        patt_quote_container = re.compile(r"{%quote_container(.*?)%}")
        patt_note_container = re.compile(r"{%note_container(.*?)%}")
        patt_tag_container = re.compile(r"{%tag_container(.*?)%}")
        output = patt_quote_container.sub(r"\1" if self.text else "", output)
        output = patt_note_container.sub(r"\1" if self.content else "", output)
        output = patt_tag_container.sub(r"\1" if self.tag else "", output)
        return pattern.sub(lambda x: replacements[x.group(0)], output)

    @property
    def colorname(self):
        """Return the stringified version of the annotation color.

        Finds the closest named color to the annotation and returns it,
        using euclidian distance between the two color vectors.
        """
        annot_colors = (
            self.colors.get("stroke") or self.colors.get("fill") or (0.0, 0.0, 0.0)
        )
        nearest = None
        minimum_similarity = COLOR_SIMILARITY_MINIMUM
        for name, values in COLORS.items():
            similarity_ratio = self._color_similarity_ratio(values, annot_colors)
            if similarity_ratio > minimum_similarity:
                minimum_similarity = similarity_ratio
                nearest = name
        return nearest

    def _color_similarity_ratio(self, color_one, color_two):
        """Return the similarity of two colors between 0 and 1.

        Takes two rgb color tuples made of floats between 0 and 1, e.g. (1, 0.65, 0) for orange,
        and returns the similarity between them, with 1 being the same color and 0 being the
        difference between full black and full white, as a float.
        """
        return 1 - (abs(math.dist([*color_one], [*color_two])) / 3)
